src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -333,40 +353,48 @@ private void InnerStartOperationReceiveFrom()
 
         internal unsafe SocketError DoOperationReceiveFrom(SafeCloseSocket handle, out SocketFlags flags)
         {
-            PrepareIOCPOperation();
-
             flags = _socketFlags;
 
-            int bytesTransferred;
-            SocketError socketError;
-            if (_buffer != null)
+            SocketError socketError = SocketError.Success;
+            NativeOverlapped* overlapped = AllocateNativeOverlapped();
+            try
             {
-                socketError = Interop.Winsock.WSARecvFrom(
-                    handle,
-                    ref _wsaBuffer,
-                    1,
-                    out bytesTransferred,
-                    ref flags,
-                    _ptrSocketAddressBuffer,
-                    _ptrSocketAddressBufferSize,
-                    _ptrNativeOverlapped,
-                    IntPtr.Zero);
+                int bytesTransferred;
+
+                if (_buffer != null)
+                {
+                    socketError = Interop.Winsock.WSARecvFrom(
+                        handle,
+                        ref _wsaBuffer,
+                        1,
+                        out bytesTransferred,
+                        ref flags,
+                        _ptrSocketAddressBuffer,
+                        _ptrSocketAddressBufferSize,
+                        overlapped,
+                        IntPtr.Zero);
+                }
+                else
+                {
+                    socketError = Interop.Winsock.WSARecvFrom(
+                        handle,
+                        _wsaBufferArray,
+                        _bufferListInternal.Count,
+                        out bytesTransferred,
+                        ref flags,
+                        _ptrSocketAddressBuffer,
+                        _ptrSocketAddressBufferSize,
+                        overlapped,
+                        IntPtr.Zero);
+                }
+
+                socketError = ProcessIOCPResult(socketError == SocketError.Success, bytesTransferred);
+                return socketError;
             }
-            else
+            finally
             {
-                socketError = Interop.Winsock.WSARecvFrom(
-                    handle,
-                    _wsaBufferArray,
-                    _bufferListInternal.Count,
-                    out bytesTransferred,
-                    ref flags,
-                    _ptrSocketAddressBuffer,
-                    _ptrSocketAddressBufferSize,
-                    _ptrNativeOverlapped,
-                    IntPtr.Zero);
+                FreeNativeOverlappedIfNotPending(overlapped, socketError);
```

Comment:
> The pattern we use in SocketPal.Windows.cs to handle the equivalent situation is that we catch, free the overlapped, and rethrow. See code below

The cases are similar but slightly different, e.g. in your example the overlapped is stored in the IAsyncResult instance, but in the SAEA case, it's only on the stack, so it would need to be passed around.  I actually initially did that, but I found it cleaner to switch to this finally model.  We can certainly revisit one or the other subsequently.